File: Algorithms/1114.按序打印/Foo_2.py
from threading import Semaphore

class Foo:
    def __init__(self):
        self.first_job_done = Semaphore(0)
        self.second_job_done = Semaphore(0)

    def first(self, printFirst: 'Callable[[], None]') -> None:
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.first_job_done.release()

    def second(self, printSecond: 'Callable[[], None]') -> None:
        self.first_job_done.acquire()
        # printSecond() outputs "second". Do not change or remove this line.
        printSecond()
        self.second_job_done.release()

    def third(self, printThird: 'Callable[[], None]') -> None:
        self.second_job_done.acquire()
        # printThird() outputs "third". Do not change or remove this line.
        printThird()

File: Concurrency/1114/Foo_2.java
class Foo {
    private AtomicBoolean firstJobDone = new AtomicBoolean(false);
    private AtomicBoolean secondJobDone = new AtomicBoolean(false);

    public void first(Runnable printFirst) throws InterruptedException {
        // printFirst.run() outputs "first". Do not change or remove this line.
        printFirst.run();
        firstJobDone.set(true);
    }

    public void second(Runnable printSecond) throws InterruptedException {
        while (!firstJobDone.get()) {
        }
        // printSecond.run() outputs "second". Do not change or remove this line.
        printSecond.run();
        secondJobDone.set(true);
    }

    public void third(Runnable printThird) throws InterruptedException {
        while (!secondJobDone.get()) {
        }
        // printThird.run() outputs "third". Do not change or remove this line.
        printThird.run();
    }
}
